Separation of feature reports across 3 devices in Win 8.1

cboulay opened the pull request:
Can someone (@nitsch) please test this in Windows 7? Or in other Windows 8 @betonme ?
All of the relevant changes are in ifdef tags so they should only affect Windows.

With these changes the controller works well for me in Windows 8.1. I can use the psmovepair utility, magnetometer_calibration, and test_calibration (USB connection required for the first time).

cboulay commented:
ugh. I was supposed to link this to #147 . Is it too late? Sorry, I'm pretty new to pull requests. Closes #147

betonme commented:
I can test it with windows 8.1 within the next days

nitsch commented:
I have test on 64-bit Windows 7. The psmoveapi test applications still
seem to work fine with your patch.

Could you add a little more detail to the comments in the code?
Something that describes the "big picture", to provide a little context
for this special treatment.

cboulay commented:
I didn't know it would delete the pull request if I deleted the branch. I should have held on to it. Sorry. It was maybe incomplete because getting the move controller working on Windows is not terribly useful without the tracker. I finally got the tracker working reliably, but I had to unlink the PS3EYEDriver from its origin repository.

Maybe I could have maintained my own fork of PS3EYEDriver then worked off that? I don't know. Sorry, I'm new to pull requests and submodules.

cboulay commented:
Oh hey I can restore and reopen it! I'll keep it open since it is a minimal fix for only the psmove controller (not the tracker). Then we can debate the merits of unlinking PS3EYEDriver another time.

cboulay commented:
@betonme , did you get a chance to test this yet?

@nitsch mentioned in a [forum post](https://groups.google.com/d/msg/psmove/sUdkR_rojt4/iBnnLf_ORmEJ) that the best Windows solution will likely be to use [Windows HID API](https://msdn.microsoft.com/en-us/library/windows/hardware/jj126202%28v=vs.85%29.aspx), which I presume will make more sense than the 3 different devices.

I'm not willing to tackle that problem, but I will bet on you to provide us with the most robust Windows solution. (Sorry, menotpunny).

thp commented on src/psmove.c:
```diff
@@ -256,6 +256,7 @@ struct _PSMove {
 
     /* The handle to the HIDAPI device */
     hid_device *handle;
+	hid_device *handle_addr; // Only used by _WIN32. Needed by Win 8.1 to get BT address.
```
Indentation

thp commented on src/psmove.c:
```diff
@@ -273,6 +274,7 @@ struct _PSMove {
 
     /* Device path of the controller */
     char *device_path;
+	char *device_path_addr;  // Only used by _WIN32. Needed by Win 8.1 to get BT address.
```
Indentation

thp commented on src/psmove.c:
```diff
@@ -505,13 +507,11 @@ psmove_count_connected_hidapi()
+         * The one with "&col02#" in the path is the one we will get the bluetooth address from.
          **/
-        if (strstr(cur_dev->path, "&col02#") == NULL) {
+        if (strstr(cur_dev->path, "&col01#") == NULL) {
```
I guess here the &col01# doesn't disappear, because we don't open it anymore?

cboulay replied:
The original comment confused me because no device was actually opened until much later, and at that point it was the device with col01 in the path that was opened. However, in this new version, I took the comment at its word and made sure to open the col02 device first (the one used for getting the BT address) and then open the col01 device.

thp commented on src/psmove.c:
```diff
@@ -560,25 +560,49 @@ psmove_connect_internal(wchar_t *serial, char *path, int id)
+	}
+	serial = NULL;  // Set the serial to NULL, even if BT device, to use psmove_get_serial below.
+    
+	/*
```
Also here in this comment block, please be consistent with indentation and other multi-line comment blocks in this file.

thp commented:
In general, this pull request has some indentation issues. Other than that looks good, but I haven't tested it on Windows.

betonme commented:
I can compile it, but the pairing is never ending.

cboulay commented:
@betonme Is that on Windows 8.1? Are you referring to the "Please unplug the USB cable and press the PS button." part? It's also never-ending for me, because there are still some changes to be made in Windows before it will pair. I suppose at this point we could check for Windows 8.1, cancel the pairing, then instruct the user to check the box in the bluetooth device services and edit the registry setting.

The better solution would be to make the device settings and registry setting changes in code, maybe restart the services, then use the Windows API for pairing (as in psmove-pair-win). That's well beyond this simple first-step.

As for the indentation, it looked OK on my screen. I'll try another editor.

betonme commented:
Yes I'm using Windows 8.1 x64. And You're right I'm referring to the "Please unplug the USB cable and press the PS button." part. It would be nice to have all the hacks done automatically in psmovepair.

cboulay commented:
I agree it would be nice.

This PR does not address the unreliability of connecting the controller in Windows 8.1 . I think that's a separate issue entirely and should probably be addressed separately.

This PR merely fixes how the API communicates with the controller once connected (either via USB or BT). Because it fixes USB communication, it makes it possible to write the host address to the controller, so the controller is now actually paired with the computer. (I bet that anyone who previously had any measure of success pairing in Windows 8.1 was using a dongle that they paired with on a different platform.) Because it can now get past the pairing phase, it exposes the connection problem. 

I think the narrow scope of this PR still warrants adding a timeout during the attempt to connect to the controller using the psmovepair utility, and if it reaches timeout to give instructions to read the README.win64 section on bluetooth connection. I don't have time to work on that until later next week.

Thanks for your feedback!

cboulay commented:
@thp , how would you like to handle the infinite connecting problem in Windows at this early stage? A timeout? A windows-only message to Ctrl+C? I know the long-term solution would be to figure out how to make all the necessary device setting + registry changes in code then restart the BT service (assuming that works), but for now, what would you like done?

thp commented:
A generous timeout (30 seconds? not sure..) would work for me, maybe with a message as well that it's buggy on Windows, will time out in %d seconds and the user can press Ctrl+C to abort early.

cboulay commented:
The latest patch includes a timeout during pairing. It builds, but I cannot test it beyond that. @betonme can you give it a shot?

nitsch commented:
> The latest patch includes a timeout during pairing.

What exactly is that timeout good for? Isn't mentioning Ctrl+C to abort
sufficient? For instance, I remember spending more than 30 seconds on
Windows 7 to get one particular Move controller connected.

Why not just display all the information prior to starting the loop and
let the user decide when to give up? It's much simpler.

betonme commented:
@cboulay Hi, it builds and here is the print out of psmovepair.exe
Connected controllers: 1
[PSMOVE WARNING] Magnetometer in 00:11:22:33:44:55 not yet calibrated.
PSMove #1 connected via USB.

    Unplug the controller.

    Now press the controller's PS button. The red status LED
    will start blinking. Whenever it goes off, press the
    PS button again. Repeat this until the status LED finally
    remains lit. This process will timeout in 30 seconds.
    Press Ctrl+C to abort early.

    A connection could not be established. This is not
    unusual in Windows. Please refer to the README document
    for your plastform for more information.
    A connection could not be established. This is not
    unusual in Windows. Please refer to the README document
    for your plastform for more information.
    A connection could not be established. This is not
    unusual in Windows. Please refer to the README document

The message continues until I hit Ctrl+C

cboulay commented:
I forgot to make it break the loop automatically, so it seems it would not have prevented @nitsch from connecting after more than 30 seconds. I clarified the text a little more. Thanks @betonme

thp commented on src/platform/psmove_winsupport.c:
```diff
@@ -311,9 +316,16 @@ windows_register_psmove(const char *move_addr_str, const HANDLE hRadio)
+            printf("\n" \
+                   "    A connection could not be established. This is not\n" \
+                   "    unusual in Windows. Please refer to the README document\n" \
+                   "    for your plastform for more information. Press Ctrl+C to cancel.");
```
Typo: plastform -> platform

thp commented on src/platform/psmove_winsupport.c:
```diff
@@ -311,9 +316,16 @@ windows_register_psmove(const char *move_addr_str, const HANDLE hRadio)
             }
         }
+        if (loop_count >= timeout_iterations) {
+            printf("\n" \
```
No need to have line continuation `\` at each end of line here. Remove those.

thp commented on src/psmove.c:
```diff
@@ -563,23 +563,47 @@ psmove_connect_internal(wchar_t *serial, char *path, int id)
         move->handle = hid_open_path(path);
-    } else {
+    }
+    else {
```
Please be consistent in that file, and place the closing brace (`}`) in the same line as the `else`.

thp commented on src/psmove.c:
```diff
@@ -599,11 +623,11 @@ psmove_connect_internal(wchar_t *serial, char *path, int id)
-
-    if (path != NULL) {
-        move->device_path = strdup(path);
+    else {
```
Same here - closing brace goes together with `else`.

thp commented on src/psmove.c:
```diff
@@ -881,8 +905,14 @@ _psmove_read_btaddrs(PSMove *move, PSMove_Data_BTAddr *host, PSMove_Data_BTAddr
+    /* _WIN32 only has move->handle_addr for getting bluetooth address. */
+    if (move->handle_addr){
+        res = hid_get_feature_report(move->handle_addr, btg, sizeof(btg));
+    }
```
`} else {`

thp commented on src/psmove.c:
```diff
@@ -994,8 +1024,13 @@ psmove_set_btaddr(PSMove *move, PSMove_Data_BTAddr *addr)
+    /* _WIN32 only has move->handle_addr for getting bluetooth address. */
+    if (move->handle_addr){
+        res = hid_send_feature_report(move->handle_addr, bts, sizeof(bts));
+    }
```
`} else {`

thp commented on src/psmove.c:
```diff
@@ -994,8 +1024,13 @@ psmove_set_btaddr(PSMove *move, PSMove_Data_BTAddr *addr)
-
-    res = hid_send_feature_report(move->handle, bts, sizeof(bts));
+    /* _WIN32 only has move->handle_addr for getting bluetooth address. */
+    if (move->handle_addr){
```
Space before opening brace:
```
if (move->handle_addr) {
```

thp commented on src/psmove.c:
```diff
@@ -1782,7 +1817,7 @@ enum PSMove_Bool
 {
     psmove_return_val_if_fail(move != NULL, 0);
-    return psmove_calibration_supported(move->calibration);
+	return psmove_calibration_supported(move->calibration);
```
Revert that indentation change. We have 4 spaces indentation.

thp commented on src/psmove.c:
```diff
@@ -1996,6 +2031,9 @@ psmove_disconnect(PSMove *move)
     switch (move->type) {
         case PSMove_HIDAPI:
             hid_close(move->handle);
+            if (move->handle_addr){// _WIN32 only
```
Space before opening brace.

thp commented on src/psmove.c:
```diff
@@ -2012,6 +2050,9 @@ psmove_disconnect(PSMove *move)
 
     free(move->serial_number);
     free(move->device_path);
+    if (move->device_path_addr){ // _WIN32 only
```
Space before opening brace.

thp commented:
Most of my comments here are about code style and indentation. I haven't tested it on Windows -- can anybody else report if it works and if it should be merged?

cboulay commented:
I squashed and rebased against master. The diff looks a bit more complicated than necessary because rebasing ended up undoing/doing indentation fixes.

thp commented:
Looks mostly good now, haven't tested it, though. Let's see if others report any issues in the coming days.

cboulay commented:
Someone that is helping me with my UE4 plugin couldn't get it to work with Windows 8.0 but it worked fine with Windows 8.1. Based on his description of the behaviour and some very simple debugging, it seems that the problem occurred when attempting to open the found BT controller, so it might be related to this part of the code. However, it also would not have worked (reliably) without these changes, so I don't think we're introducing any regressions. Further, I don't intend to support Windows 8.0. The upgrade to 8.1 is free and 8.1 is significantly better in many aspects.

At some point it might be necessary to use platform-specific communication API (hidapi in Linux, OSX, and Windows' bluetooth API in Windows). My guess is that this would solve a lot of problems with pairing and communication.
